File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/DateRangePicker.java
package com.gb.canibuythat.ui;

import android.app.DatePickerDialog;
import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RadioGroup;

import com.gb.canibuythat.R;
import com.gb.canibuythat.util.DateUtils;

import java.time.LocalDate;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DateRangePicker extends LinearLayout {

    public interface TouchInterceptor {
        boolean onInterceptTouchEvent(MotionEvent ev);
    }

    @BindView(R.id.start_date_btn)
    Button startDateBtn;
    @BindView(R.id.end_date_btn)
    Button endDateBtn;
    @BindView(R.id.reset_btn)
    ImageView resetBtn;

    private LocalDate startDate;
    private LocalDate endDate;

    private TouchInterceptor touchInterceptor;

    private DatePickerDialog.OnDateSetListener dateSetListener = new DatePickerDialog.OnDateSetListener() {
        @Override
        public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
            LocalDate newDate = LocalDate.of(year, month, dayOfMonth);

            switch ((int) view.getTag()) {
                case R.id.start_date_btn:
                    startDate = newDate;
                    startDateBtn.setText(DateUtils.formatDayMonthYearWithPrefix(newDate));
                    if (endDate.isBefore(newDate)) {
                        endDatePickerDialog = new DatePickerDialog(getContext(), dateSetListener, newDate.getYear(), newDate.getMonthValue(), newDate.getDayOfMonth());
                        setEndDate(newDate);
                    }
                    startDateChanged = true;
                    break;
                case R.id.end_date_btn:
                    endDate = newDate;
                    endDateBtn.setText(DateUtils.formatDayMonthYearWithPrefix(newDate));
                    if (startDate.isAfter(newDate)) {
                        startDatePickerDialog = new DatePickerDialog(getContext(), dateSetListener, newDate.getYear(), newDate.getMonthValue(), newDate.getDayOfMonth());
                        setStartDate(newDate);
                    }
                    endDateChanged = true;
                    break;
            }
        }
    };
    private View.OnClickListener datePickerOnClickListener = v -> {
        switch (v.getId()) {
            case R.id.start_date_btn:
                getStartDatePickerDialog().show();
                startDateBtn.setError(null);
                break;
            case R.id.end_date_btn:
                getEndDatePickerDialog().show();
                endDateBtn.setError(null);
                break;
        }
    };
    private boolean startDateChanged;
    private boolean endDateChanged;
    private DatePickerDialog startDatePickerDialog;
    private DatePickerDialog endDatePickerDialog;

    public DateRangePicker(Context context) {
        super(context);
        init();
    }

    public DateRangePicker(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.date_range_picker, this);
        setOrientation(VERTICAL);
        ButterKnife.bind(this);
        resetDates();
        RadioGroup radioGroup = new RadioGroup(getContext());

        startDateBtn.setOnClickListener(datePickerOnClickListener);
        endDateBtn.setOnClickListener(datePickerOnClickListener);
        resetBtn.setOnClickListener(v -> {
            resetDates();
            startDatePickerDialog = null;
            endDatePickerDialog = null;
        });
        setOrientation(LinearLayout.HORIZONTAL);
        setGravity(Gravity.CENTER_VERTICAL);
    }

    public void setStartDate(LocalDate startDate) {
        this.startDate = startDate;
        updateButtons();
        startDatePickerDialog = null;
    }

    public void setEndDate(LocalDate endDate) {
        this.endDate = endDate;
        updateButtons();
        endDatePickerDialog = null;
    }

    public LocalDate getStartDate() {
        return startDate;
    }

    public LocalDate getEndDate() {
        return endDate;
    }

    public boolean isStartDateChanged() {
        return startDateChanged;
    }

    public boolean isEndDateChanged() {
        return endDateChanged;
    }

    private void resetDates() {
        startDate = LocalDate.now();
        endDate = LocalDate.now();
        updateButtons();
    }

    private void updateButtons() {
        startDateBtn.setText(DateUtils.formatDayMonthYearWithPrefix(this.startDate));
        endDateBtn.setText(DateUtils.formatDayMonthYearWithPrefix(this.endDate));
    }

    private DatePickerDialog getStartDatePickerDialog() {
        if (startDatePickerDialog == null) {
            startDatePickerDialog = new DatePickerDialog(getContext(), dateSetListener,
                    startDate.getYear(), startDate.getMonthValue(), startDate.getDayOfMonth());
        }
        startDatePickerDialog.getDatePicker().setTag(R.id.start_date_btn);
        return startDatePickerDialog;
    }

    private DatePickerDialog getEndDatePickerDialog() {
        if (endDatePickerDialog == null) {
            endDatePickerDialog = new DatePickerDialog(getContext(), dateSetListener,
                    endDate.getYear(), endDate.getMonthValue(), endDate.getDayOfMonth());
        }
        endDatePickerDialog.getDatePicker().setTag(R.id.end_date_btn);
        return endDatePickerDialog;
    }

    public void setTouchInterceptor(TouchInterceptor touchInterceptor) {
        this.touchInterceptor = touchInterceptor;
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        if (touchInterceptor != null) {
            return touchInterceptor.onInterceptTouchEvent(ev);
        } else {
            return super.onInterceptTouchEvent(ev);
        }
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/DateRangePicker.kt
package com.gb.canibuythat.ui

import android.app.DatePickerDialog
import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.DatePicker
import android.widget.LinearLayout
import butterknife.ButterKnife
import com.gb.canibuythat.R
import com.gb.canibuythat.util.createDatePickerDialog
import com.gb.canibuythat.util.formatDayMonthYear
import kotlinx.android.synthetic.main.date_range_picker.view.*
import java.time.LocalDate

class DateRangePicker : LinearLayout {

    var startDate: LocalDate = LocalDate.now()
        set(value) {
            field = value
            start_date_btn.text = value.formatDayMonthYear()
            startDatePickerDialog = null
        }

    var endDate: LocalDate = LocalDate.now()
        set(value) {
            field = value
            end_date_btn.text = value.formatDayMonthYear()
            endDatePickerDialog = null
        }

    var touchInterceptor: ((ev: MotionEvent) -> Boolean)? = null

    private val onDateSetListener: (DatePicker, LocalDate) -> Unit = { datePicker, date ->
        when (datePicker.tag as Int) {
            R.id.start_date_btn -> {
                startDate = date
                if (endDate < date) {
                    endDate = date
                }
                isStartDateChanged = true
            }
            R.id.end_date_btn -> {
                endDate = date
                if (startDate > date) {
                    startDate = date
                }
                isEndDateChanged = true
            }
        }
    }

    private val datePickerOnClickListener = { v: View ->
        when (v.id) {
            R.id.start_date_btn -> {
                getStartDatePickerDialog().show()
                start_date_btn.error = null
            }
            R.id.end_date_btn -> {
                getEndDatePickerDialog().show()
                end_date_btn.error = null
            }
        }
    }
    var isStartDateChanged: Boolean = false
        private set
    var isEndDateChanged: Boolean = false
        private set
    private var startDatePickerDialog: DatePickerDialog? = null
    private var endDatePickerDialog: DatePickerDialog? = null

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    private fun init() {
        View.inflate(context, R.layout.date_range_picker, this)
        orientation = LinearLayout.VERTICAL
        ButterKnife.bind(this)
        resetDates()

        start_date_btn.setOnClickListener(datePickerOnClickListener)
        end_date_btn.setOnClickListener(datePickerOnClickListener)
        reset_btn.setOnClickListener { v ->
            resetDates()
            startDatePickerDialog = null
            endDatePickerDialog = null
        }
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL
    }

    private fun resetDates() {
        startDate = LocalDate.now()
        start_date_btn.text = startDate.formatDayMonthYear()
        endDate = LocalDate.now()
        end_date_btn.text = endDate.formatDayMonthYear()
    }

    private fun getStartDatePickerDialog(): DatePickerDialog {
        startDatePickerDialog = startDatePickerDialog ?: createDatePickerDialog(context, startDate, onDateSetListener).apply {
            datePicker.tag = R.id.start_date_btn
        }
        return startDatePickerDialog!!
    }

    private fun getEndDatePickerDialog(): DatePickerDialog {
        endDatePickerDialog = endDatePickerDialog ?: createDatePickerDialog(context, endDate, onDateSetListener).apply {
            datePicker.tag = R.id.end_date_btn
        }
        return endDatePickerDialog!!
    }

    override fun onInterceptTouchEvent(ev: MotionEvent) = touchInterceptor?.invoke(ev)
            ?: super.onInterceptTouchEvent(ev)
}
